tests/xharness/TestReporter.cs

```diff
@@ -0,0 +1,490 @@
+		{
+			runner = appRunner ?? throw new ArgumentNullException (nameof (appRunner));
+			deviceName = device; // can be null on simulators 
+			listener = simpleListener ?? throw new ArgumentNullException (nameof (simpleListener));
+			runLog = log ?? throw new ArgumentNullException (nameof (log));
+			crashLogs = new Logs (runner.Logs.Directory);
+			crashReporter = crashReports ?? throw new ArgumentNullException (nameof (crashReports));
+			Timeout = runner.GetNewTimeout ();
+			TimeoutWatch  = new Stopwatch ();
+			CallbackLog = new CallbackLog ((line) => {
+				// MT1111: Application launched successfully, but it's not possible to wait for the app to exit as requested because it's not possible to detect app termination when launching using gdbserver
+				waitedForExit &= line?.Contains ("MT1111: ") != true;
+				if (line?.Contains ("error MT1007") == true)
+					launchFailure = true;
+			});
+		}
+
+		// parse the run log and decide if we managed to start the process or not
+		async Task<(int pid, bool launchFailure)> GetPidFromRunLog () {
+			(int pid, bool launchFailure) pidData = (-1, true);
+			using var reader = runLog.GetReader (); // diposed at the end of the method, which is what we want
+			if (reader.Peek () == -1) {
+				// empty file! we definetly had a launch error in this case
+				pidData.launchFailure = true;
+			} else {
+				while (!reader.EndOfStream) {
+					var line = await reader.ReadLineAsync ();
+					if (line.StartsWith ("Application launched. PID = ", StringComparison.Ordinal)) {
+						var pidstr = line.Substring ("Application launched. PID = ".Length);
+						if (!int.TryParse (pidstr, out pidData.pid))
+							runner.MainLog.WriteLine ("Could not parse pid: {0}", pidstr);
+					} else if (line.Contains ("Xamarin.Hosting: Launched ") && line.Contains (" with pid ")) {
+						var pidstr = line.Substring (line.LastIndexOf (' '));
+						if (!int.TryParse (pidstr, out pidData.pid))
+							runner.MainLog.WriteLine ("Could not parse pid: {0}", pidstr);
+					} else if (line.Contains ("error MT1008")) {
+						pidData.launchFailure = true;
+					}
+				}
+			}
+			return pidData;
+		}
+
+		// parse the main log to get the pid 
+		async Task<int> GetPidFromMainLog ()
+		{
+			int pid = -1;
+			using var log_reader = runner.MainLog.GetReader (); // dispose when we leave the method, which is what we want
+			string line;
+			while ((line = await log_reader.ReadLineAsync ()) != null) {
+				const string str = "was launched with pid '";
+				var idx = line.IndexOf (str, StringComparison.Ordinal);
+				if (idx > 0) {
+					idx += str.Length;
+					var next_idx = line.IndexOf ('\'', idx);
+					if (next_idx > idx)
+						int.TryParse (line.Substring (idx, next_idx - idx), out pid);
+				}
+				if (pid != -1)
+					return pid;
+			}
+			return pid;
+		}
+
+		// return the reason for a crash found in a log
+		void GetCrashReason (int pid, ILog crashLog, out string crashReason)
```

Comment:
Same code as we used to have inside the RunAsync method, just moved it out so that is easier to understand what is happening.